tespy/components/customs.py

```diff
@@ -0,0 +1,818 @@
+        if self.overheating.val is False:
+            o3 = self.outl[2].to_flow()
+            self.mat_deriv[k, 5, 1] = -dh_mix_dpQ(o3, 1)
+            self.mat_deriv[k, 5, 2] = 1
+            k += 1
+
+    def fluid_func(self):
+        r"""
+        Calculates the vector of residual values for component's fluid balance
+        equations.
+
+        Returns
+        -------
+        vec_res : list
+            Vector of residual values for component's fluid balance.
+
+            .. math::
+
+                0 = fluid_{i,in_{j}} - fluid_{i,out_{j}} \;
+                \forall i \in \mathrm{fluid}, \; \forall j \in inlets/outlets
+        """
+        vec_res = []
+
+        for i in range(self.num_i):
+            for fluid, x in self.inl[i].fluid.val.items():
+                vec_res += [x - self.outl[i].fluid.val[fluid]]
+        return vec_res
+
+    def mass_flow_func(self):
+        r"""
+        Calculates the residual value for component's mass flow balance
+        equation.
+
+        Returns
+        -------
+        vec_res : list
+            Vector with residual value for component's mass flow balance.
+
+            .. math::
+
+                0 = \dot{m}_{in,i} - \dot{m}_{out,i} \;
+                \forall i \in inlets/outlets
+        """
+        vec_res = []
+        for i in range(self.num_i):
+            vec_res += [self.inl[i].m.val_SI - self.outl[i].m.val_SI]
+        return vec_res
+
+    def fluid_deriv(self):
+        r"""
+        Calculates the partial derivatives for all fluid balance equations.
+
+        Returns
+        -------
+        deriv : list
+            Matrix with partial derivatives for the fluid equations.
+        """
+        deriv = np.zeros((self.num_nw_fluids * 3,
+                          6 + self.num_vars,
+                          self.num_nw_vars))
+        # hot side 1
+        i = 0
+        for fluid in self.nw_fluids:
+            deriv[i, 0, i + 3] = 1
+            deriv[i, 3, i + 3] = -1
+            i += 1
+        # hot side 2
+        j = 0
+        for fluid in self.nw_fluids:
+            deriv[i + j, 1, j + 3] = 1
+            deriv[i + j, 4, j + 3] = -1
+            j += 1
+        # cold side
+        k = 0
+        for fluid in self.nw_fluids:
+            deriv[i + j + k, 2, k + 3] = 1
+            deriv[i + j + k, 5, k + 3] = -1
+            k += 1
+        return deriv.tolist()
+
+    def mass_flow_deriv(self):
+        r"""
+        Calculates the partial derivatives for all mass flow balance equations.
+
+        Returns
+        -------
+        deriv : list
+            Matrix with partial derivatives for the mass flow balance
+            equations.
+        """
+        deriv = np.zeros((3, 6 + self.num_vars, self.num_nw_vars))
+        for i in range(self.num_i):
+            deriv[i, i, 0] = 1
+        for j in range(self.num_o):
+            deriv[j, j + i + 1, 0] = -1
+        return deriv.tolist()
```

Comment:
The same as above (sh)(c)ould be applied here.

Reply:
Changed.